File: src/helpers/oauth2-helper.ts
#!/usr/bin/env node

/**
 * OAuth2トークン取得ヘルパースクリプト
 *
 * このスクリプトは、Google Sheets APIのOAuth2認証用のトークンを取得します。
 */

import { OAuth2Client } from "google-auth-library";
import http from "http";
import { URL } from "url";
import path from "path";
import { fileURLToPath } from "url";
import open from "open";
import readline from "readline";

const SCOPES = ["https://www.googleapis.com/auth/spreadsheets.readonly"];
const REDIRECT_URI = "http://localhost:3000/oauth2callback";

interface OAuth2Credentials {
  clientId: string;
  clientSecret: string;
}

/**
 * ユーザーにクライアントIDとシークレットを入力させる
 */
async function promptForCredentials(): Promise<OAuth2Credentials> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) => {
    console.log("\n📋 Google Cloud ConsoleからOAuth 2.0クライアントの情報を入力してください\n");

    rl.question("Client ID: ", (clientId) => {
      rl.question("Client Secret: ", (clientSecret) => {
        rl.close();
        resolve({ clientId, clientSecret });
      });
    });
  });
}

/**
 * ローカルサーバーを起動して認証コードを受け取る
 */
async function startLocalServer(oauth2Client: OAuth2Client): Promise<string> {
  return new Promise((resolve, reject) => {
    const server = http.createServer(async (req, res) => {
      try {
        if (req.url && req.url.indexOf("/oauth2callback") > -1) {
          const qs = new URL(req.url, REDIRECT_URI).searchParams;
          const code = qs.get("code");

          if (!code) {
            res.end("認証に失敗しました: コードが取得できませんでした");
            reject(new Error("No code found"));
            return;
          }

          res.end("認証成功! このタブを閉じて、ターミナルに戻ってください。");
          server.close();
          resolve(code);
        }
      } catch (e) {
        reject(e);
      }
    });

    server.listen(3000, () => {
      console.log("\n🌐 ローカルサーバーを起動しました (http://localhost:3000)");
    });
  });
}

/**
 * OAuth2認証フローを実行してトークンを取得
 */
async function getTokens(oauth2Client: OAuth2Client): Promise<void> {
  // 認証URLを生成
  const authorizeUrl = oauth2Client.generateAuthUrl({
    access_type: "offline",
    scope: SCOPES,
    prompt: "consent", // 常にリフレッシュトークンを取得
  });

  console.log("\n🔐 認証URLを開いています...");
  console.log("   ブラウザが自動的に開かない場合は、以下のURLを手動で開いてください:");
  console.log(`   ${authorizeUrl}\n`);

  // ブラウザで認証URLを開く
  try {
    await open(authorizeUrl);
  } catch {
    console.warn("⚠️  ブラウザを自動的に開けませんでした");
  }

  // ローカルサーバーで認証コードを受け取る
  const code = await startLocalServer(oauth2Client);

  console.log("\n✅ 認証コードを取得しました");
  console.log("🔄 トークンを取得中...\n");

  // トークンを取得
  const { tokens } = await oauth2Client.getToken(code);

  console.log("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
  console.log("✅ トークンの取得に成功しました!");
  console.log("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n");

  console.log("📝 以下の情報を設定ファイルに追加してください:\n");
  console.log("oauth2:");
  console.log(`  clientId: ${oauth2Client._clientId}`);
  console.log(`  clientSecret: ${oauth2Client._clientSecret}`);
  console.log(`  redirectUri: ${REDIRECT_URI}`);
  console.log(`  refreshToken: ${tokens.refresh_token}`);
  if (tokens.access_token) {
    console.log(
      `  accessToken: ${tokens.access_token}  # オプション（有効期限: ${tokens.expiry_date ? new Date(tokens.expiry_date).toLocaleString() : "不明"}）`
    );
  }

  console.log("\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n");

  if (!tokens.refresh_token) {
    console.warn("⚠️  警告: リフレッシュトークンが取得できませんでした");
    console.warn("   これは、以前に同じクライアントIDで認証した可能性があります。");
    console.warn("   対処方法:");
    console.warn("   1. Google アカウントの「アプリのアクセス権」からアプリを削除");
    console.warn("   2. このスクリプトを再実行");
    console.warn("   または、generateAuthUrl の prompt オプションを確認してください。\n");
  }
}

/**
 * メイン処理
 */
export async function runOauth2Helper() {
  console.log("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
  console.log("🔑 OAuth2 トークン取得ヘルパー");
  console.log("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");

  try {
    // クライアントIDとシークレットを取得
    const { clientId, clientSecret } = await promptForCredentials();

    if (!clientId || !clientSecret) {
      console.error("\n❌ Client IDとClient Secretは必須です");
      process.exit(1);
    }

    // OAuth2クライアントを初期化
    const oauth2Client = new OAuth2Client({
      clientId,
      clientSecret,
      redirectUri: REDIRECT_URI,
    });

    // トークンを取得
    await getTokens(oauth2Client);

    console.log("💡 次のステップ:");
    console.log("   1. 上記の設定を test-oauth2.config.yaml にコピー");
    console.log("   2. path にスプレッドシートIDを設定");
    console.log("   3. node lib/cli.js --config test-oauth2.config.yaml\n");
  } catch (error) {
    console.error("\n❌ エラーが発生しました:", (error as Error).message);
    process.exit(1);
  }
}

const executedPath = process.argv[1] ? path.resolve(process.argv[1]) : "";
const currentPath = fileURLToPath(import.meta.url);

// スクリプトとして実行された場合のみ実行
if (executedPath && executedPath === currentPath) {
  runOauth2Helper().catch((error) => {
    console.error("\n❌ エラーが発生しました:", (error as Error).message);
    process.exit(1);
  });
}

File: rust/l10n-rust/src/oauth2_helper.rs
use std::io::{self, Write};

use anyhow::{Context, Result, anyhow, bail};
use tiny_http::{Response, Server, StatusCode};
use url::Url;

use crate::google::extract_oauth_error_message;
use crate::model::OAuthTokenResponse;

pub(crate) fn run_oauth2_helper() -> Result<()> {
    println!("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
    println!("🔑 OAuth2 トークン取得ヘルパー (Rust)");
    println!("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n");

    let client_id = prompt_input("Client ID: ")?;
    let client_secret = prompt_input("Client Secret: ")?;
    if client_id.trim().is_empty() || client_secret.trim().is_empty() {
        bail!("Client IDとClient Secretは必須です");
    }

    let redirect_uri = "http://localhost:3000/oauth2callback";
    let auth_url = build_google_oauth_authorize_url(&client_id, redirect_uri)?;

    println!("以下のURLをブラウザで開いて認証してください:");
    println!("{auth_url}\n");

    println!("ローカルサーバーを起動して認証コードを待機中... (http://localhost:3000)");
    let code = wait_for_oauth_code()?;
    println!("✅ 認証コードを取得しました");

    let token = exchange_oauth_authorization_code(&client_id, &client_secret, redirect_uri, &code)?;

    println!("\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
    println!("✅ トークンの取得に成功しました!");
    println!("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n");
    println!("oauth2:");
    println!("  clientId: {client_id}");
    println!("  clientSecret: {client_secret}");
    println!("  redirectUri: {redirect_uri}");
    if let Some(refresh_token) = token.refresh_token.as_deref() {
        println!("  refreshToken: {refresh_token}");
    } else {
        println!("  # refreshToken: 未取得（prompt=consent や権限再付与を確認してください）");
    }
    println!("  accessToken: {}", token.access_token);
    if let Some(expires_in) = token.expires_in {
        println!("  # expiresIn: {expires_in} seconds");
    }

    Ok(())
}

fn build_google_oauth_authorize_url(client_id: &str, redirect_uri: &str) -> Result<String> {
    let mut url = Url::parse("https://accounts.google.com/o/oauth2/v2/auth")
        .with_context(|| "Failed to build OAuth authorize URL")?;
    url.query_pairs_mut()
        .append_pair("response_type", "code")
        .append_pair("client_id", client_id)
        .append_pair("redirect_uri", redirect_uri)
        .append_pair(
            "scope",
            "https://www.googleapis.com/auth/spreadsheets.readonly",
        )
        .append_pair("access_type", "offline")
        .append_pair("prompt", "consent");
    Ok(url.to_string())
}

fn wait_for_oauth_code() -> Result<String> {
    let server = Server::http("127.0.0.1:3000")
        .map_err(|error| anyhow!("Failed to start local server on :3000: {error}"))?;

    for request in server.incoming_requests() {
        let request_url: &str = request.url();
        let callback_url = format!("http://localhost{request_url}");
        if let Some(code) = parse_code_from_callback_url(&callback_url) {
            let response =
                Response::from_string("認証成功! このタブを閉じて、ターミナルに戻ってください。")
                    .with_status_code(StatusCode(200));
            let _ = request.respond(response);
            return Ok(code);
        }

        let response =
            Response::from_string("Not Found. /oauth2callback?code=... を呼び出してください。")
                .with_status_code(StatusCode(404));
        let _ = request.respond(response);
    }

    bail!("OAuth2認証コードを取得できませんでした")
}

fn parse_code_from_callback_url(callback_url: &str) -> Option<String> {
    let url = Url::parse(callback_url).ok()?;
    if url.path() != "/oauth2callback" {
        return None;
    }
    url.query_pairs()
        .find_map(|(key, value)| (key == "code").then(|| value.to_string()))
}

fn exchange_oauth_authorization_code(
    client_id: &str,
    client_secret: &str,
    redirect_uri: &str,
    code: &str,
) -> Result<OAuthTokenResponse> {
    let token_endpoint = "https://oauth2.googleapis.com/token";
    let response = reqwest::blocking::Client::new()
        .post(token_endpoint)
        .form(&[
            ("grant_type", "authorization_code"),
            ("code", code),
            ("client_id", client_id),
            ("client_secret", client_secret),
            ("redirect_uri", redirect_uri),
        ])
        .send()
        .with_context(|| "Failed to exchange OAuth2 authorization code")?;

    if !response.status().is_success() {
        let status_text = response.status().to_string();
        let body = response.text().unwrap_or_default();
        let message = extract_oauth_error_message(&body).unwrap_or(status_text);
        bail!("{message}");
    }

    response
        .json::<OAuthTokenResponse>()
        .with_context(|| "Failed to parse OAuth2 code exchange response")
}

fn prompt_input(message: &str) -> Result<String> {
    print!("{message}");
    io::stdout()
        .flush()
        .with_context(|| "Failed to flush stdout")?;
    let mut input = String::new();
    io::stdin()
        .read_line(&mut input)
        .with_context(|| "Failed to read stdin")?;
    Ok(input.trim().to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_code_from_callback_url_extracts_code() {
        let code =
            parse_code_from_callback_url("http://localhost/oauth2callback?code=abc123&scope=test");
        assert_eq!(code, Some("abc123".to_string()));
    }
}
